subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java

```diff
@@ -122,38 +121,55 @@ public void setReportsDir(File reportsDir) {
-        extension.setDestinationFile(providerFactory.provider(new Callable<File>() {
+        extension.setDestinationFile(providers.provider(new Callable<File>() {
             @Override
             public File call() {
-                return fileOperations.file(projectLayout.getBuildDirectory().get().getAsFile() + "/jacoco/" + taskName + ".exec");
+                return layout.getBuildDirectory().file("jacoco/" + taskName + ".exec").get().getAsFile();
             }
         }));
 
         task.getJvmArgumentProviders().add(new JacocoAgent(extension));
-        task.doFirst(new Action<Task>() {
-            @Override
-            public void execute(Task task) {
-                if (extension.isEnabled() && extension.getOutput() == JacocoTaskExtension.Output.FILE) {
-                    // Delete the coverage file before the task executes, so we don't append to a leftover file from the last execution.
-                    // This makes the task cacheable even if multiple JVMs write to same destination file, e.g. when executing tests in parallel.
-                    // The JaCoCo agent supports writing in parallel to the same file, see https://github.com/jacoco/jacoco/pull/52.
-                    File coverageFile = extension.getDestinationFile();
-                    if (coverageFile == null) {
-                        throw new GradleException("JaCoCo destination file must not be null if output type is FILE");
-                    }
-                    fileOperations.delete(coverageFile);
-                }
-            }
-        });
+        task.doFirst(new JacocoOutputCleanupTestTaskAction(
+            fs,
+            providers.provider(() -> extension.isEnabled() && extension.getOutput() == JacocoTaskExtension.Output.FILE),
+            providers.provider(extension::getDestinationFile)
+        ));
 
         // Do not cache the task if we are not writing execution data to a file
-        task.getOutputs().doNotCacheIf("JaCoCo configured to not produce its output as a file", new Spec<Task>() {
-            @Override
-            public boolean isSatisfiedBy(Task element) {
-                // Do not cache Test task if Jacoco doesn't produce its output as files
-                return extension.isEnabled() && extension.getOutput() != JacocoTaskExtension.Output.FILE;
+        Provider<Boolean> doNotCachePredicate = providers.provider(() ->
+            // Do not cache Test task if Jacoco doesn't produce its output as files
+            extension.isEnabled() && extension.getOutput() != JacocoTaskExtension.Output.FILE
+        );
+        task.getOutputs().doNotCacheIf(
+            "JaCoCo configured to not produce its output as a file",
+            element -> doNotCachePredicate.get()
+        );
+    }
+
+    private static class JacocoOutputCleanupTestTaskAction implements Action<Task> {
```

Comment:
Using a serializable lambda here didn't work.

![image](https://user-images.githubusercontent.com/132773/94273383-a2a80200-ff44-11ea-9f27-f8bf30d6c0b8.png)

Reply:
It seems it was capturing `this` (the extension) at that point. Maybe a variable declaration was missing?

Reply:
There was a reference to an outer field for `FileSystemOperations`. Once assigned to a local var the configuration cache is happy (and intellij says the var is redundant of course).

But then the up_to_dateness / caching behavior of the task is broken by the lambda. I'll leave it with the static class.